### PR TITLE
chore: improves API usability.

### PR DESCRIPTION
**Pull request by jcchavezs**
This PR breaks the API (already broken when we moved to PHP 7) and notably we changed the way we build spans:

From
```php
$span = $tracer->newChild($context);
$span->setName('encode');
$span->start(1554988060000);
try {
  doSomethingExpensive();
} finally {
  $span->finish();
}
```

to 

```php
$span = $tracer->startSpan('encode', ['parent' => $context, 'start_time' => 1554988060000]);
try {
  doSomethingExpensive();
} finally {
  $span->finish();
}
```

or if we want to create a child on the current one:

```php
$span = $tracer->startNextSpan('encode', ['start_time' => 1554988060000]);
try {
  doSomethingExpensive();
} finally {
  $span->finish();
}
```

This API is inspired in brave (which zipkin PHP v1 was based on) and zipkin-go which has many usability tweaks. Changes are:

- Spans are meant to be started instead of being created and then started (like zipkin-go does)
- Span methods are fluent.

Ping @adriancole @basvanbeek @ghermeto @telemmaite @anensshiya-govinthas-ck

### CODE REVIEW

File: src/Zipkin/NoopSpan.php [hunk 1]
```diff
@@ -51,20 +51,22 @@ public function getContext(): TraceContext
      * set its name without lock contention.
      *
      * @param int|null $timestamp
-     * @return void
+     * @return self
      */
-    public function start(?int $timestamp = null): void
+    public function start(?int $timestamp = null): self
```
Comment (jcchavezs) [inline src/Zipkin/NoopSpan.php]: Making it fluent might not be necessary if we allow to pass tags on creation and start the span by default.

File: src/Zipkin/Tracer.php [hunk 2]
```diff
@@ -81,28 +81,73 @@ public function __construct(
      * }</pre>
      *
      * @param SamplingFlags $samplingFlags
+     * @param array $options
      * @return Span
      */
-    public function newTrace(SamplingFlags $samplingFlags = null): Span
+    public function startSpan(string $name, ?array $options = []): Span
```
Comment (jcchavezs) [inline src/Zipkin/Tracer.php]: first new method. Conflates newTrace and newChild.

File: src/Zipkin/Tracer.php [hunk 3]
```diff
@@ -81,28 +81,73 @@ public function __construct(
...
+            $span->setKind($options['kind']);
+        }
+        
+        return $span->setName($name)->start($options['start_time'] ?? null);
     }
 
-    /**
-     * Creates a new span within an existing trace. If there is no existing trace, use {@link
-     * #newTrace()} instead.
+        /**
```
Comment (jcchavezs) [inline src/Zipkin/Tracer.php]: Don't pay attention to this comments yet.

File: src/Zipkin/Tracer.php [hunk 4]
```diff
@@ -81,28 +81,73 @@ public function __construct(
...
+     * so that data is written to the correct trace.
      *
-     * @param TraceContext $parent
+     * @param SamplingFlags|TraceContext $contextOrFlags
      * @return Span
-     * @throws \RuntimeException
+     * @throws RuntimeException
      */
-    public function newChild(TraceContext $parent): Span
+    public function startNextSpan(string $name, ?array $options = []): Span
```
Comment (jcchavezs) [inline src/Zipkin/Tracer.php]: replaces `nextSpan`

**Comment by adriancole** [comment]
I presume the started right away thing is for local tracing. functionality for deferred start may still come in useful for some instrumentation, though possibly not in php

**Comment by jcchavezs** [comment]
As for the instrumentation I wrote, I can't recall having this need but
could you please give me some example use cases on this? The only case
comes to my mind is in zipkin-instrumentation-sql where we defer the span
creation based on the output of a method (but we use the start time of when
the method was originally called).

Thanks!

fre. 12. apr. 2019, 00:26 skrev Adrian Cole <notifications@github.com>:

**Comment by adriancole** [comment]
mostly we use span.start deferred because you want the timing to be about the execution of the call. if do things like tagging prior to span.start, you can get an accurate measurement of what the call took vs time to derive tags. Another thing is sometimes there's a deferred start in async instrumentation. For example, the difference between assembling a call and invoking it

File: src/Zipkin/NoopSpan.php [hunk 5]
```diff
@@ -111,10 +116,11 @@ public function annotate(string $value, int $timestamp): void
      * It is often expensive to derive a remote address: always check {@link #isNoop()} first!
      *
      * @param Endpoint $remoteEndpoint
-     * @return void
+     * @return self
      */
-    public function setRemoteEndpoint(Endpoint $remoteEndpoint): void
+    public function setRemoteEndpoint(Endpoint $remoteEndpoint): Span
```
Comment (adriancole) [inline src/Zipkin/NoopSpan.php]: maybe remove set as others do?

File: src/Zipkin/RealSpan.php [hunk 6]
```diff
@@ -32,15 +32,15 @@ private function __construct(TraceContext $context, Recorder $recorder)
      * @param Recorder $recorder
      * @return RealSpan
      */
-    public static function create(TraceContext $context, Recorder $recorder): self
+    public static function create(TraceContext $context, Recorder $recorder): Span
```
Comment (adriancole) [inline src/Zipkin/RealSpan.php]: interesting that a utility like this is public? does it need to be? or is it too late to change

File: src/Zipkin/RealSpan.php [hunk 7]
```diff
@@ -64,10 +64,10 @@ public function getContext(): TraceContext
      * set its name without lock contention.
      *
      * @param int $timestamp
-     * @return void
+     * @return self
      * @throws \InvalidArgumentException
      */
-    public function start(?int $timestamp = null): void
```
Comment (adriancole) [inline src/Zipkin/RealSpan.php]: maybe mention if called later this overwrites the timestamp

File: src/Zipkin/RealSpan.php [hunk 8]
```diff
@@ -80,17 +80,19 @@ public function start(?int $timestamp = null): void
...
 
     /**
      * Sets the string name for the logical operation this span represents.
      *
      * @param string $name
-     * @return void
+     * @return self
      */
-    public function setName(string $name): void
+    public function setName(string $name): Span
```
Comment (adriancole) [inline src/Zipkin/RealSpan.php]: you remove setTag for tag earlier.. wondering if you aim to be consistent all over or just above

File: src/Zipkin/Tracer.php [hunk 9]
```diff
@@ -76,55 +70,122 @@ public function __construct(
...
-     *   return $this->tracer->newTrace($flags);
+     *   return $this->tracer->startSpan('request', ['parent' => $flags]);
      * }
      * }</pre>
-     *
-     * @param SamplingFlags $samplingFlags
-     * @return Span
      */
-    public function newTrace(SamplingFlags $samplingFlags = null): Span
+    public function startSpan(string $name, array $options = []): Span
```
Comment (adriancole) [inline src/Zipkin/Tracer.php]: in brave this is special cased for also applying scope eg startScopedSpan. I'm not sure how important it is to save the call to start. Seems higher value doing more than this to break api. I don't know what features are needed, but the rationale for scoped span was arbitrary tracing. advanced span (ex including deferred start) is for instrumentation and usually any repetition is hidden with shared abstractions like http.

**Pull request closed without merging**